x-pack/plugins/security_solution/server/lib/telemetry/tasks/indices.metadata.ts

```diff
@@ -0,0 +1,198 @@
+      const trace = taskMetricsService.start(taskType);
+
+      const taskConfig = telemetryConfiguration.indices_metadata_config;
+
+      // TODO: not use taskExecutionPeriod, it's just to test the task using the temporary API
+      const queryConfig: QueryConfig = {
+        maxPrefixes: Number(taskExecutionPeriod.last ?? taskConfig.max_prefixes),
+        maxGroupSize: Number(taskExecutionPeriod.current ?? taskConfig.max_group_size),
+      };
+
+      const publishClusterStats = (stats: ClusterStats) => {
+        sender.reportEBT(TELEMETRY_CLUSTER_STATS_EVENT.eventType, stats);
+      };
+
+      const publishDatastreamsStats = (stats: DataStream[]): number => {
+        let counter = 0;
+        for (const ds of stats) {
+          sender.reportEBT(TELEMETRY_DATA_STREAM_EVENT.eventType, ds);
+          counter++;
+        }
+        log.info(`Sent data streams`, { count: counter } as LogMeta);
+        return counter;
+      };
+
+      const publishIndicesStats = async (indices: string[]): Promise<number> => {
+        let counter = 0;
+        for await (const stat of receiver.getIndicesStats(
+          indices.slice(0, taskConfig.indices_threshold),
+          queryConfig
+        )) {
+          sender.reportEBT(TELEMETRY_INDEX_STATS_EVENT.eventType, stat);
+          counter++;
+        }
+        log.info(`Sent indices stats`, { count: counter } as LogMeta);
+        return counter;
+      };
+
+      const publishIlmStats = async (indices: string[]): Promise<Set<string>> => {
+        const ilmNames = new Set<string>();
+        for await (const stat of receiver.getIlmsStats(indices, queryConfig)) {
+          if (stat.policy_name !== undefined) {
+            ilmNames.add(stat.policy_name);
+            sender.reportEBT(TELEMETRY_ILM_STATS_EVENT.eventType, stat);
+          }
+        }
+        log.info(`Sent ILM stats`, { count: ilmNames.size } as LogMeta);
+
+        return ilmNames;
+      };
+
+      const publishIlmPolicies = async (ilmNames: Set<string>): Promise<number> => {
+        let counter = 0;
+        for await (const policy of receiver.getIlmsPolicies(
+          Array.from(ilmNames.values()),
+          queryConfig
+        )) {
+          sender.reportEBT(TELEMETRY_ILM_POLICY_EVENT.eventType, policy);
+          counter++;
+        }
+        log.info(`Sent ILM policies`, { count: counter } as LogMeta);
+        return counter;
+      };
+
+      const incrementCounter = (type: TelemetryCounter, name: string, value: number) => {
+        const telemetryUsageCounter = sender.getTelemetryUsageCluster();
+        telemetryUsageCounter?.incrementCounter({
+          counterName: createUsageCounterLabel(COUNTER_LABELS.concat(name)),
+          counterType: type,
+          incrementBy: value,
+        });
+      };
+
+      try {
+        // 1. Get cluster stats and list of indices and datastreams
+        const [clusterStats, indices, dataStreams] = await Promise.all([
+          receiver.getClusterStats(),
```

Comment:
Should we perform the cluster stats call separately first to early exit if count of indices is too high? Just thinking of a scenario where `getIndices` and/or `getDataStreams` cause any memory issues for super clusters

Reply:
Yes, it was the initial implementation, but after doing the maths:

- clusterStats: `ClusterStats`  -> **40 bytes**
- indices: `string[]` -> supposing 20K indices with 50 chars length on average -> **2MB**
- dataStreams: `DataStream[]` -> supposing 1K datastreams, each datastream having 500 indices and a 50 chars lenght name: **~4MB**

```js
export interface ClusterStats {
  num_nodes: number;
  num_indices: number;
  num_docs: number;
  num_deleted_docs?: number;
  total_size_in_bytes: number;
}
                                
export interface Index {
  index_name: string;
  ilm_policy?: string;
}
                                
export interface DataStream {
  datastream_name: string;
  indices?: Index[];
}
```
Since for a high number of indices and datastreams the memory footprint was ~6MB, I changed it.

Once we have the initial list of indices and datastreams, we apply the config to take the first N elements for both lists and then "manually paginate" the stats queries.

I think we are safe with the memory needed for this first execution, even for super clusters, wdyt?